bugfix: GetFinalCommandLines(string baseCommandLine) now also works for test-containers with a huge number of test-suites

MartinOSchmidt opened the pull request:
Dear Christian,
if one has a very large number of test-suites in a test.exe (e.g. hundreds or thousands of suites), and tries to use trait-filters, chances are high that CommandLineGenerator.GetFinalCommandLines() fails to work properly due to the code line

    int remainingLength = MaxCommandLength
    - baseCommandLineWithFilter.Length - _lengthOfExecutableString - userParam.Length - 1;

where remainingLength may become negative.
Find attached a patch-file (only the changes to CommandLineGenerator.cs are of interest) to fix this problem by modifying the used approach.
Would be great if the changes could be incorporated into the master/trunk.

Feel free to ask anything with respect to my changes.

Cheers

MartinOSchmidt commented:
no clue what all these errors are about ... ??? local build just works fine

csoltenborn commented:
You are probably on a pre-V15.4 version of VS - MS has broken integration with the `VsTest.Console.exe` for my adapter. I just have fixed the build - if there are any more problems, they should be on you now ;-)

You can try this by discarding the changes to the three non-`.cs` files, which contain changes produced by the setup script that must not be commited (I know that this is unfortunate - will be fixed in the future).

MartinOSchmidt commented:
I'm on VS Pro 15.4.5 and the  appveyor-errors also occured without the "local changes"-commit induced by "build_preparation.bat" (which I pushed after the first ".cs-file only" push)

MartinOSchmidt commented:
Luckily your build-fix solved the problem :) 👍

csoltenborn commented:
Oops, I meant pre-V15.5... Ok, looks good so far, give me some time for the code review... Could you maybe in the meantime add tests to `CommandLineGeneratorTests` which demonstrate the bug (i.e., would fail without your patch, but pass now)?

Aren't pull requests awesome? And we haven't even done the code review :-)

MartinOSchmidt commented:
I'll do as soon as I find the time ...

And yes, I like it :)

MartinOSchmidt commented:
Added test to demonstrate the bug and also fixed an issue with my bugfix
(which occured with the new test which is a even worse scenario as we had here ;) )

MartinOSchmidt commented:
Hi Christian, do you already know when you'll find the time to do the code review (and hopefully pull the changes)?

csoltenborn commented:
Sorry for not coming back to you earlier... Let's start the review right now - I will have a look at the test code (and hopefully review the rest at the weekend). Thanks already for your contribution!

csoltenborn commented on GoogleTestAdapter/Core/Runners/CommandLineGenerator.cs:
```diff
@@ -69,39 +69,59 @@ private IEnumerable<Args> GetFinalCommandLines(string baseCommandLine)
+            int groupIdxCounter = 0;
+            int maxSuiteLength = MaxCommandLength - _lengthOfExecutableString - userParam.Length - 1;
+
+            var suiteGroups = suitesRunningAllTests
```
This is some neat usage of Linq :-) However, I feel that we are kind of mixing abstraction levels here. I would recommend moving this code (i.e., lines 74 to 87?) into an appropriately named method.

csoltenborn commented on GoogleTestAdapter/Core/Runners/CommandLineGenerator.cs:
```diff
@@ -69,39 +69,59 @@ private IEnumerable<Args> GetFinalCommandLines(string baseCommandLine)
+
+            var suiteGroups = suitesRunningAllTests
+                              .Select(suite =>
+                                { suiteLengthAggregate += suite.Length + 3;
```
I had to think about the 3 - probably not a good sign. Maybe use a `const int separatorLength = ".*;".Length`?

MartinOSchmidt replied:
yep!

csoltenborn commented on GoogleTestAdapter/Core/Runners/CommandLineGenerator.cs:
```diff
@@ -69,39 +69,59 @@ private IEnumerable<Args> GetFinalCommandLines(string baseCommandLine)
+                              .GroupBy(pair => pair.groupIdx, pair => pair.suiteName)
+                              .ToList();
+
+            if (!suiteGroups.Any()) suiteGroups = Enumerable.Range(-1, 1).GroupBy(dummy => dummy, dummy => string.Empty).ToList();
```
What's this? Don't we just need to proceed to test case handling if we do not have any suiteGroups, i.e., we do not have any suites at all?

MartinOSchmidt replied:
This is only a trick so I don't need no (or nearly no) special handling for this case but just can iterater "normally" over the suiteGroups (i.e. without an entry in suiteGroups nothing would be processed).

csoltenborn replied:
Ok, I see - my confusion was caused by assuming that the `Enumerable.Range()` parameters were lower and upper bound (not all of us happen to be Linq wizards ;-) ). However, I think there's a better option:
* just collect the suite groups as you already do
* iterate over the first `n-1` of them, basically performing your `if` case
* then perform your `else` case (outside the loop)

There's some tiny code duplication from the pre-`if` part of the loop (which could be reduced e.g. by introducing method `GetTestCasesOfSuiteGroup()`, which might indeed be a good idea), but I think it's easier to understand (especially after extracting the method as suggested at line 77, since no knowledge of your trick is necessary at the call side).

MartinOSchmidt replied:
Your proposed approach would not help in avoiding the 'trick', because the trick is not about special handling for the last groupSuite but about having a groupSuite at all. The trick would be much more elegant if i could create the 'fake'-grouping which results in an empty string-list if ToList() is applied on therefore the `suiteGroup.Key >= 0` clause would be unneccessary (but I failed to do so and I did not want to spend to much time on this).

And yes I could seperate things as you propose, but I would need to extract the "preamble" (before the `if (suiteGroup != suiteGroups.Last())`) into a method (or lambda) to avoid code-duplication. Not sure if this improves readability/coherence a lot, but I will give it a try (however not before monday - because I don't have the right VisualStudio version installed here)

csoltenborn commented on GoogleTestAdapter/Core/Runners/CommandLineGenerator.cs:
```diff
@@ -69,39 +69,59 @@ private IEnumerable<Args> GetFinalCommandLines(string baseCommandLine)
-            while (testCasesNotRunBySuite.Count > 0)
-            {
-                remainingLength = MaxCommandLength
+              List<string> suites = suiteGroup.Key >= 0 ? suiteGroup.ToList() : Enumerable.Empty<string>().ToList();
```
I also do not get this, but it's probably related to the above...

MartinOSchmidt replied:
yes. the "little bit" of special handling needed. As I only have a "fake" suiteGroup (if Key < 0) I need to construct a empty list of suites (as suiteGroup.ToList() would not work properly here).

MartinOSchmidt replied:
So I moved the suite-list-creation into its own method, removed all the "special handling", introduced the "preamble" as a Func (with anonymous return type), and split the loop into the "all but the last suite-list" part and the last suite-list part (which also creates the command lines with non-suite-filters)
Hopefully this improved readability and coherence.

csoltenborn commented on GoogleTestAdapter/Core/Runners/CommandLineGenerator.cs:
```diff
@@ -210,20 +230,27 @@ private bool AllTestCasesOfExecutableAreRun()
-                }
-            }
-            return testCasesNotRunBySuite;
+            return _testCasesToRun.Where(t => !suitesRunningAllTests.Contains(GetTestsuiteName(t))).ToList();
```
That's nice! Feel free to remove the other code...

MartinOSchmidt replied:
OK :)

csoltenborn commented on GoogleTestAdapter/Core.Tests/Runners/CommandLineGeneratorTests.cs:
```diff
@@ -309,6 +309,27 @@ public void GetCommandLines_ManyTestsWithSuites_BreaksUpLongCommandLinesCorrectl
+      List<CommandLineGenerator.Args> commands = new CommandLineGenerator(testCases, TestDataCreator.DummyExecutable.Length, "", "", TestEnvironment.Options)
+          .GetCommandLines().ToList();
+
+      commands.Count.Should().Be(4);
```
I think it would make sense to make sure that the resulting commands are not shorter than `maxLength - lengthOfLongestSuiteName`. Otherwise, we might need more invocations than actually necessary, increasing execution time (and this would at least be a reasonable heuristics, I guess).

MartinOSchmidt replied:
Sound reasonable!

csoltenborn commented on GoogleTestAdapter/Core.Tests/Runners/CommandLineGeneratorTests.cs:
```diff
@@ -309,6 +309,27 @@ public void GetCommandLines_ManyTestsWithSuites_BreaksUpLongCommandLinesCorrectl
 
+    [TestMethod]
+    [TestCategory(Unit)]
+    public void GetCommandLines_HugeNumberOfSuitesToBeExecutedFully_WontBreakCommandLineCreation()
```
Since this is absolutely crucial code for GTA, I feel we should add some more test cases (e.g., with/without user parameters; different execution cases, e.g. with/without a test not run by a suite (I think the "no suite" case should be covered by the other tests))... What do you think?

MartinOSchmidt replied:
What you propose sounds reasonable too me. However I don't have an overview about the current state of test-coverage. What you propose is, if I understand this correctly, independent of my changes?
I wrote the test (as you originally proposed) with the intention to show that without my changes the method will throw an exception (and not more).

csoltenborn replied:
If you mean test coverage in the technical sense (which you probably don't), you can check this out via the CodeCov links (and it should also be shown on GitHub - if not, let me know). Reproducing the bug with a test is important, but since you have changed handling of suites right in the core of GTA, we should also make sure that the new code behaves correctly (and will not be broken in the future). However, ... (see below)

csoltenborn commented on GoogleTestAdapter/Core.Tests/Runners/CommandLineGeneratorTests.cs:
```diff
@@ -309,6 +309,27 @@ public void GetCommandLines_ManyTestsWithSuites_BreaksUpLongCommandLinesCorrectl
+
+      IEnumerable<Model.TestCase> testCases = TestDataCreator.CreateDummyTestCasesFull(testsToExecute.ToArray(), allTests.ToArray());
+
+      List<CommandLineGenerator.Args> commands = new CommandLineGenerator(testCases, TestDataCreator.DummyExecutable.Length, "", "", TestEnvironment.Options)
```
Since the xml file parameter will never be empty as far as I recall, it's probably better to pass a non-empty string for that parameter.

MartinOSchmidt replied:
I just did copy & paste from one of the existing test ... you're the master here ;)

csoltenborn replied:
I was pretty sure that I would complain about stuff that I had done myself (and which you had used as base of your code) :-) :-)

MartinOSchmidt replied:
:D - isn't is terrible that if you come back to your code half a year later you sometimes think - who implemented this crap ;)

MartinOSchmidt commented on GoogleTestAdapter/Core.Tests/Runners/CommandLineGeneratorTests.cs:
```diff
@@ -309,6 +309,27 @@ public void GetCommandLines_ManyTestsWithSuites_BreaksUpLongCommandLinesCorrectl
+    {
+      var range = Enumerable.Range(0, 400);
+
+      var allTests = range.Select(idx => "MyTestSuiteWithSomeWhatLongNameIsStillHandledCorrectly" + idx + ".Test")
```
You're right about that it is a little bit complicated. I made the test-suite-collection a little bit more complex to check the changed algorithm returns the right result set also for somewhat complicated collections (which I partially check by 
`commands.First().TestCases.Count(testCase => testCase.DisplayName.EndsWith("Test3")).Should().Be(1);`).
I create a test-suite-collection & test-set-to-execute, where some suites are fully tested, some are partially tested and where only a single suite with a 'Test3' is executed.
However for purely testing on the bug this is not neccessary (and maybe placed better in seperate tests) and your proposed test-set might be sufficient.

csoltenborn replied:
I just realized that the most important corner cases of your code (e.g. "no suites", "one resulting command line only with suites", "one resulting command with suites and test cases") should be covered by the existing tests. Therefore, your test is probably fine - maybe add a tiny comment giving the rational for how you construct the test cases, and add some more assertions...

MartinOSchmidt replied:
I thought so. As I did not change the behavior or API there should be no need for any additional tests (except those that have been missing beforehand). Actually as I made my test more complicate than neccessary for reproducing the bug, I found a bug in my original bugfix (which was not found by any of the existing tests) ... still it might be a good idea to seperate concerns ...

MartinOSchmidt replied:
Finally improved documentation and added some addtional asserts :)

MartinOSchmidt commented:
So i hopefully did all the requested changes.

csoltenborn commented on GoogleTestAdapter/Core/Runners/CommandLineGenerator.cs:
```diff
@@ -59,6 +59,8 @@ public IEnumerable<Args> GetCommandLines()
             return commandLines;
         }
 
+        private Func<A, R> GetFuncWithAnonymousReturnType<A, R>(Func<A, R> f) { return f; }
```
This is an interesting approach - have learned something here...

Due to the generality of this method, I suggest moving it to a new class `Common.Utils`.

csoltenborn commented:
Looks just great now, and I have also learned something :-) I have add one final request for change which you might even appreciate. Will merge afterwards...

csoltenborn commented:
Just wanted to let you know that I can of course do the refactoring on my own - just thought you might like the idea to move some of your code to some general place. Are you going to do the change, or shall I merge and then do it myself?

MartinOSchmidt commented:
sorry, i did not come back earlier. First I had no time and then I was struck down by the influenza - could not do anything last week. I will move the code as soon as I feel well enough (hopefully the next days).

csoltenborn commented:
Ok, don't worry, and get better soon!

MartinOSchmidt commented:
So I moved the auxiliary function GetFuncWithAnonymousReturnType()  to the new static class GoogleTestAdapter.Common.Functional (I don't like "Utils"-classes that much) and renamed it to ToFunc() (and added some documentation).
If you anyhow prefer it to be in a "Utils" class, feel free to rename the thing (but don't forget to change the documentation ;) )

MartinOSchmidt commented:
Did you have a look?

csoltenborn commented:
Now I have, sorry for the delay - the code looks just fine, but I fear I still have one question: Why the Microsoft copyright statement at the beginning of `Functional.cs`? Are you contributing on behalf of MS?

MartinOSchmidt commented:
Haha, for sure not! :)
I wondered myself, where these lines were coming from. But as other files in the project also had these comments (e.g. String.cs, IString.cs and IDebuggerAttacherService.cs), I thought this was some kind of solution-setting and did not care too much ...

MartinOSchmidt commented:
Now I remember, what I did: I first added the new class to String.cs to check if everything works as expected and than I applied a move-refactoring (moving the class to its own file). I assume the refactoring just copied the comment-lines ...
I will remove those lines ...

csoltenborn commented:
Ok, I see :-) The files you mentioned have indeed be added by MS, but in general, my repository is licensed under Apache (as you probably know). Thanks for removing them!

csoltenborn commented:
Ok, merging is done - thanks once more for your contribution (and your patience ;-) )...

I will release a new version soon (in the next weeks at the latest).

MartinOSchmidt commented:
Great! And thanks for the support in my first contribution to github :) 👍
